File: 1079-letter-tile-possibilities/1079-letter-tile-possibilities.java
class Solution {
void tile(String tiles,HashSet<String>forcount,StringBuilder temp,HashSet<Integer>idx){
    if(temp.length()>tiles.length()){
        return ;
    }
    if(!temp.isEmpty()&&!forcount.contains(temp.toString())){
        forcount.add(temp.toString());
    }
    for(int i=0;i<tiles.length();i++){
        if(!idx.contains(i)){
            temp.append(tiles.charAt(i));
            idx.add(i);
            tile(tiles,forcount,temp,idx);
            temp.deleteCharAt(temp.length()-1);
            idx.remove(i);
        }
    }
}
    public int numTilePossibilities(String tiles) {
        HashSet<String>forcount=new HashSet<>();
        HashSet<Integer>idx=new HashSet<>();
        StringBuilder temp=new StringBuilder("");
        tile(tiles,forcount,temp,idx);
        return forcount.size();
    }
}

File: 1079-letter-tile-possibilities/1079-letter-tile-possibilities.cpp
class Solution {
    void tile(string tiles, unordered_set<string>& forcount, string& temp, unordered_set<int>& idx) {
        if (temp.length() > tiles.length()) {
            return;
        }
        if (!temp.empty() && forcount.find(temp) == forcount.end()) {
            forcount.insert(temp);
        }
        for (int i = 0; i < tiles.length(); i++) {
            if (idx.find(i) == idx.end()) {
                temp.push_back(tiles[i]);
                idx.insert(i);
                tile(tiles, forcount, temp, idx);
                temp.pop_back(); // backtrack
                idx.erase(i);
            }
        }
    }

public:
    int numTilePossibilities(string tiles) {
        unordered_set<string> forcount;
        unordered_set<int> idx;
        string temp = "";
        tile(tiles, forcount, temp, idx);
        return forcount.size();
    }
};